### ISSUE TITLE
Inline Pages Structure With Dynamic Navbar

### ISSUE THREAD

**Issue by JasonKleban** [bencompton/framework7-react#54]
I cannot find the framework7-react equivalent documentation to (Inline Pages Structure With Dynamic Navbar)[http://framework7.io/docs/pages-inline.html#inline-pages-structure-with-dynamic-navbar].  As far as I can tell, I'm having to hack like this?

```
<View main navbarThrough dynamicNavbar domCache>
    <div className="navbar">
        <div className="navbar-inner" data-page="Today">
            <NavLeft><a href="#" className="link icon-only open-panel"><i className="icon icon-bars"></i></a></NavLeft>
            <NavCenter>Today</NavCenter>
        </div>
    </div>
</View>
```

This also ends up rendering an extra `<div class="navbar"></div>` above my explicit one.  Is there a way to make <Navbar /> render to a `.navbar-inner` all wrapped in the `.navbar` when this `dynamicNavbar domCache` sitation is detected?

**Comment by JasonKleban** [comment]
This also becomes a problem once I start updating the react props causing the components to re-render and override f7's `class="page cached"` for example.  How is this correctly dealt with?  I can't find any discussion of it.

**Comment by thiagoabreu** [comment]
I think I may help you but I need some info first:

1- What are you trying to get specifically? The dynamic navbar or the
inline pages ( "DOM cache")?
2- If it is the navbar, is your target iOS or Android?

Em dom, 16 de jul de 2017 10:29, JasonKleban <notifications@github.com>
escreveu:

**Comment by MrAndMrsK** [comment]
Great!

1) both.  I want the navbar with the sliding transitions and I want to use the domcache inline pages so that all the content is on the page at once and just being switched dynamically, calls to a content server.  But react app state updates too, which messes with f7's css-class modifications.

2) ios

**Comment by JasonKleban** [comment]
Great!

1) both.  I want the navbar with the sliding transitions and I want to use the domcache inline pages so that all the content is on the page at once and just being switched dynamically, calls to a content server.  But react app state updates too, which messes with f7's css-class modifications.

2) ios

**Comment by JasonKleban** [comment]
Does framework7-react support domCache mode?

**Comment by thiagoabreu** [comment]
This is the closest I could get to the "inline pages" example in the F7 docs.

```jsx
const App = (props, context) => (
  <Framework7App themeType="ios" routes={[]}>
    <Statusbar />
    <Views>
      <View id="main-view" navbarThrough dynamicNavbar main url="/" domCache>
        <Navbar afterInnerSlot={
          <div>
            <div className="navbar-inner cached" data-page="about">
              <div className="center">About</div>
            </div>
            <div className="navbar-inner cached" data-page="services">
              <div className="center">Services</div>
            </div>
          </div>
        }>
          <NavCenter>Home</NavCenter>
        </Navbar>
        <Pages>
          <Page name="home">
            <p>Page 1</p>
            <p>
              <Link pageName="about" href="#about" text="go to about" />
            </p>
            <p>
              <Link pageName="services" href="#services" text="go to services" />
            </p>
          </Page>
          <Page name="about" cached>
            <Link icon="icon-back" href="#home" back />
            <p>Page 2</p>
          </Page>
          <Page name="services" cached>
            <Link icon="icon-back" href="#home" back />
            <p>Page 3</p>
          </Page>
        </Pages>
      </View>
    </Views>
  </Framework7App>
)
```

BTW, in the [View docs for the framework7-react](https://framework7.io/react/view.html) is said that the `domCached` prop is enabled by default.

**Comment by JasonKleban** [comment]
Thank you for this example.  I think I have implemented these concepts too, so that's good confirmation. 

But how does framework7 toggling the `cached` class on elements in the dom jibe with the re-rendering by the react components whenever the props change?  

This example will work until a prop changes, triggering a re-render, overwriting the modified dom elements back to the initial configuration.  Right?

**Comment by bencompton** [comment]
> BTW, in the View docs for the framework7-react is said that the domCached prop is enabled by default.

This is correct; DOM caching is the default and most well-tested mode in framework7-react. The kitchen sink and starter templates demonstrate how to use DOM caching.

>But how does framework7 toggling the cached class on elements in the dom jibe with the re-rendering by the react components whenever the props change?

Yes, good question. React ends up rendering new Page components into the DOM (via code from Framework7 Vue), and then Framework7 gets called to perform the animation and class toggling on the Navbar and other Page components. When React re-renders pages, the page divs of course do not get removed and re-added, and instead get various DOM patches applied. React's reconciler has no knowledge of the CSS classes that Framework7 added, and therefore does not create DOM patches to remove them.

React and Vue docs of course do not recommended having external JavaScript code modifying DOM elements that React / Vue control, but in this case we're deliberately breaking the rules. The reason is because the page animation, namely iOS page animation, is quite finicky and complex and would be difficult to accomplish with pure React / Vue code (I know, I tried). This approach, while perhaps not the greatest example of technical correctness, has worked very well in both React and Vue.

**Comment by JasonKleban** [comment]
Note: this was driving me nuts ... when using the `class` representation of Components, we MUST specify contextTypes with a `'framework7AppContext'` keyed function or else it will get stripped off (except in react production environments).

```
export class App extends Component {
    static contextTypes = {
        framework7AppContext: () => {}
    }
```

Sure, this should be PropTypes.function but I'm actually using typescript so I don't want to bother with run-time type checking or the added dependency of PropTypes - this no-op function seems to do the trick and have no react errors.

**Comment by JasonKleban** [comment]
But there's still something wrong here ...  Here's the LeftPanel of the [Framework7 React Starter Template](http://framework7.io/react/templates.html) converted to use `class` syntax so that I have access to things like `componentDidMount()`.

```
class LeftPanel extends Component {
    static contextTypes = {
        framework7AppContext: () => {}
    }

    render() {
        return <Panel left reveal layout="dark">
            <View id="left-panel-view" navbarThrough dynamicNavbar="true">
                {this.context.framework7AppContext.theme.ios ? <Navbar title="Left Panel"></Navbar> : null}
                <Pages>
                    <Page>
                        {this.context.framework7AppContext.theme.material ? <Navbar title="Left Panel"></Navbar> : null}
                        <ContentBlock inner>
                            <p>Left panel content goes here</p>
                        </ContentBlock>
                        <ContentBlockTitle>Load page in panel</ContentBlockTitle>
                        <List>
                            <ListItem link="/about/" title="About"></ListItem>
                            <ListItem link="/form/" title="Form"></ListItem>
                        </List>
                        <ContentBlockTitle>Load page in main view</ContentBlockTitle>
                        <List>
                            <ListItem link="/about/" title="About" linkView="#main-view" linkClosePanel></ListItem>
                            <ListItem link="/form/" title="Form" linkView="#main-view" linkClosePanel></ListItem>
                        </List>
                    </Page>
                </Pages>
            </View>
        </Panel>;
    }

    componentDidMount() {
        console.log(this.context.framework7AppContext.getRouter());
    }
}
```

Because I added `contextTypes`, `this.context.framework7AppContext` IS still defined in context.  `getRouter()`, however, returns `undefined`.

**Comment by thiagoabreu** [comment]
Try to add a constructor to your class. Something like this: 

```ts
contructor(props:any, context:any) { 
     super(props, context) 
     // ...
}
```

**Comment by JasonKleban** [comment]
@thiagoabreu - that isn't making a difference for me.  Is it for you?
